File: kotlin/src/main/kotlin/3Sum.kt
fun threeSum(nums: IntArray): List<List<Int>> {
    val arrayList = ArrayList<List<Int>>()
    nums.sort()
    for (i in 0 until nums.size) {
        if (i > 0 && nums[i] == nums[i - 1]) {
            continue
        }
        val want = 0 - nums[i]
        var l = i + 1
        var r = nums.size - 1
        while (l < r && l >= i + 1 && r <= nums.size - 1) {
            if (l > i + 1 && nums[l] == nums[l - 1]) {
                l++
                continue
            }
            if (r < nums.size - 1 && nums[r] == nums[r + 1]) {
                r--
                continue
            }
            when {
                nums[l] + nums[r] < want -> l++
                nums[l] + nums[r] == want -> {
                    arrayList.add(listOf(nums[i], nums[l], nums[r]))
                    l++
                    r--
                }
                nums[l] + nums[r] > want -> r--
            }
        }
    }
    return arrayList
}

File: golang/3Sum.go
package main

import (
	"fmt"
	"sort"
)

func threeSum(nums []int) [][]int {
	sort.Ints(nums)
	result := make([][]int, 0)
	for i := 0; i < len(nums); {
		needSum := 0 - nums[i]
		for l, r := i+1, len(nums)-1; l < r; {
			if nums[l]+nums[r] < needSum {
				l++
			} else if nums[l]+nums[r] > needSum {
				r--
			} else {
				result = append(result, []int{nums[i], nums[l], nums[r]})
				for l++; l < r && nums[l] == nums[l-1]; l++ {
				}
				for r--; l < r && nums[r] == nums[r+1]; r-- {
				}
			}
		}
		for i++; i < len(nums)-1 && nums[i] == nums[i-1]; i++ {
		}
	}
	return result
}

func main() {
	fmt.Println(threeSum([]int{-1, 0, 1, 2, -1, -4}))
}
